Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopDetector.swift

```diff
@@ -0,0 +1,180 @@
+        
+        loadEvents()
+        addStartupEvent()
+        saveEvents()
+
+        consecutiveCrashesCount = consecutiveEventsCount()
+
+        /*
+            true -> crash loop detected -> set safe mode
+            false -> crash loop NOT detected -> set normal mode
+         */
+        let result = consecutiveCrashesCount >= BacktraceCrashLoopDetector.consecutiveCrashesThreshold
+        BacktraceCrashLoop.LogDebug("Finishing Crash Loop Detection: \(result)")
+        return result
+    }
+
+    @objc private func loadEvents() {
+        
+        // Cleanup old events - f.e. for multiple usages of detector
+        startupEvents.removeAll()
+        
+        /*
+         - Since detector's DB is relatively small, UserDefaults are a good option here,
+         plus they allow to avoid a headache with reading/writing to/from the custom file.
+
+         - But we should consider shared computers as well - comment from UserDefaults docs:
+            With the exception of managed devices in educational institutions,
+            a user’s defaults are stored locally on a single device,
+            and persisted for backup and restore.
+            To synchronize preferences and other data across a user’s connected devices,
+            use NSUbiquitousKeyValueStore instead.
+         */
+        guard let data = UserDefaults.standard.object(forKey: BacktraceCrashLoopDetector.plistKey) as? Data
+        else { return }
+
+        guard let array = try? PropertyListDecoder().decode([StartUpEvent].self, from: data)
+        else { return }
+
+        startupEvents.append(contentsOf: array)
+        BacktraceCrashLoopDetector.LogDebug("Events Loaded: \(startupEvents.count)")
+    }
+
+    @objc internal func saveEvents() {
+        let data = try? PropertyListEncoder().encode(startupEvents)
+        UserDefaults.standard.set(data, forKey: BacktraceCrashLoopDetector.plistKey)
+        BacktraceCrashLoopDetector.LogDebug("Events Saved: \(startupEvents.count)")
+    }
+
+    @objc internal func addStartupEvent() {
+        
+        let crashesCount = BacktraceCrashLoopCounter.crashesCount()
+        let event = StartUpEvent(uuid: UUID().uuidString,
+                                 timestamp: Double(Date.timeIntervalSinceReferenceDate),
+                                 crashesCount: crashesCount)
+        
+        BacktraceCrashLoop.LogDebug(event.description())
+
+        startupEvents.append(event)
+        
+        if startupEvents.count > BacktraceCrashLoopDetector.consecutiveCrashesThreshold {
+            startupEvents.removeFirst()
+        }
+
+        BacktraceCrashLoop.LogDebug("Startup Event Added: \(startupEvents.count)")
+    }
+
+    @objc internal func clearStartupEvents() {
+        startupEvents.removeAll()
+        saveEvents()
+        BacktraceCrashLoop.LogDebug("Startup Events Cleared: \(startupEvents.count)")
+    }
+    
+    @objc internal func consecutiveEventsCount() -> Int {
+        
+        var count = 0
```

Comment:
The addStartupEvent drops the first event if the number of events is greater than the counter. You might have a situation when you will have the following number of events with crashCount property
[4,5,1,2,3] for threshold 5

let's see what is the result of the algorithm for the example above
4, 5 is going to increase counter to 2
then 1,2,3 won't increase the counter.

so as you can see the counter should be equal to 3, but is equal to 2 instead. Not sure if I'm missing something but - there is a high chance to have this situation in production code. ALso the question related to the way how we increase counter and when we reset it is really relevant to this question